# **Python String Formatting**

String formatting in Python allows developers to embed variables, calculations, or logic into strings, creating dynamic and readable messages.

## **What is String Formatting?**
String formatting:
- Combines variables and text dynamically.
- Supports calculations, function calls, and logic inline.
- Ensures precise formatting (e.g., setting decimal places or aligning text).

**Benefits:**
- Improves readability and reduces complexity.
- Avoids manual concatenation.
- Provides precise control over the formatting of data.


## **1. String Formatting with the `%` Operator**
The `%` operator is an older method for string formatting, using placeholders to represent variables.

### **How It Works**
- **Placeholders**:
  - `%s`: For strings.
  - `%d`: For integers.
  - `%f`: For floating-point numbers.
- **Modifiers**:
  - `%.2f`: Limits floating-point numbers to 2 decimal places.
  - `%x`: Converts numbers to hexadecimal.
  - `%r`: Inserts the "representation" of an object, providing a detailed and unambiguous view, useful for debugging.

### **Explanation of the Formatting:**

- **`%s`**: Inserts the string `"Alice"`.
- **`%d`**: Inserts the integer `25`.
- **`%.1f`**: Formats the floating-point number `5.6` with **1 decimal place**.
- **`%.2f`**: Formats the floating-point number `19.99` with **2 decimal places**.
- **`%x`**: Converts the integer `255` into **hexadecimal** format (`ff`).
- **`%r`**: Inserts the **representation** of the `data` list (`[1, 2, 3]`).







#### **Example 1: Floating-Point Numbers**


In [48]:
price = 19.99
quantity = 3

# Format the message with placeholders:
# %.2f - Shows the price with 2 decimal places
# %d - Shows the quantity as an integer
message = "The price per item is £%.2f and I bought %d items." % (price, quantity)
print(message)


The price per item is £19.99 and I bought 3 items.


### Explanation:
- **`%.2f`**: Placeholder for floating-point values, formatted to 2 decimal places (e.g., `19.99`).
- **`%d`**: Placeholder for integer values (e.g., `3`).




#### **Example 2: Hexadecimal Numbers**


In [49]:
number = 255

# Formatting the number using placeholders:
# %d - Displays the number in decimal format (e.g., 255).
# %x - Displays the number in hexadecimal format (e.g., ff).

# Create a formatted message
message = "The number %d in hexadecimal is %x." % (number, number)

# Print the formatted message
print(message)


The number 255 in hexadecimal is ff.


### Explanation:
- **`%d`**: Converts the number to a decimal (base 10) representation.
- **`%x`**: Converts the number to a lowercase hexadecimal (base 16) representation.



#### **Example 3: Mixed Data Types**


In [50]:
name = "KB"  # Name as a string
age = 30     # Age as an integer
height = 5.6 # Height as a floating-point number

# Format a message combining different data types:
# %s - Inserts the string (name).
# %d - Inserts the integer (age).
# %.1f - Inserts the floating-point number (height) rounded to 1 decimal place.

message = "Name: %s, Age: %d, Height: %.1f feet" % (name, age, height)

# Print the formatted message
print(message)


Name: KB, Age: 30, Height: 5.6 feet


### Explanation:
- **`%s`**: Used to include the string value of `name`.
- **`%d`**: Used to include the integer value of `age`.
- **`%.1f`**: Formats the floating-point value of `height` to show 1 decimal place.



#### **Example 4: Using `%r` for Any Python Object**


In [51]:
data = [1, 2, 3]  # A list of numbers

# %r shows the raw representation of the object
message = "The data list is: %r" % data

# Print the message
print(message)


The data list is: [1, 2, 3]


### Explanation:
- **`%r`**: Represents any Python object as a string. This is useful for debugging because it shows the raw representation of the object.
- The `data` list `[1, 2, 3]` is directly converted into its string form and included in the message.



# **2. F-Strings: The Modern Approach**

### **Key Features**
1. **Definition**:
   - An f-string is a string prefixed with `f` or `F`.
   - Placeholders `{}` evaluate and insert variables, expressions, or function results.

   Example:
   ```python
   name = "Alice"
   greeting = f"Hello, {name}!"
   print(greeting)
   ```
   **Output**:
   ```
   Hello, Alice!
   ```

2. **Advantages**:
   - **Readable**: Simplified syntax compared to older methods.
   - **Dynamic**: Supports inline calculations, logic, and function calls.
   - **Efficient**: Faster processing at runtime.

### **Examples**

#### **Step 1: Basic Variable Insertion**


In [52]:
name = "KB"  # Define the name variable
age = 30     # Define the age variable

# Create a dynamic message using f-strings
# f-strings allow embedding variables directly in the string using {variable_name}
message = f"Hello, my name is {name} and I am {age} years old."

# Print the dynamic message
print(message)


Hello, my name is KB and I am 30 years old.


### Explanation:
- **f-string**: A formatted string literal (introduced in Python 3.6) that allows you to embed variables or expressions inside curly braces `{}`.
- **Dynamic message**: The values of `name` and `age` are directly included in the string without the need for concatenation or type conversion.
- **Output**: The message dynamically reflects the values of the variables:


#### **Step 2: Inline Calculations**
Performing arithmetic inside an f-string:

In [53]:
# Define the price of one item
price = 20

# Define the quantity of items purchased
quantity = 3

# Calculate the total cost directly in the f-string
# {price * quantity} performs the multiplication to calculate total cost
# :.2f formats the result to 2 decimal places (e.g., 60.00)
message = f"The total cost is ${price * quantity:.2f}."

# Print the formatted message
print(message)


The total cost is $60.00.


### Explanation:
- **Inline calculation**: The expression `price * quantity` is directly embedded in the f-string and evaluated within the string.
- **:.2f**: This modifier formats the result to display two decimal places (e.g., `60.00`).
- **Output**: The result dynamically calculates the total cost and displays it in the format:


#### **Step 3: Combining Variables and Text**


In [54]:
# Define the name, city, and age variables
name = "Bob"
city = "New York"
age = 28

# Combine variables into a structured message using f-strings
# {name}, {city}, and {age} are directly embedded into the string
message = f"Hi, my name is {name}. I live in {city}, and I am {age} years old."

# Print the message
print(message)


Hi, my name is Bob. I live in New York, and I am 28 years old.


#### **Step 4: Formatting Percentages**

Convert a decimal value into a percentage:


In [55]:
# Define the success rate as a decimal
success_rate = 0.875

# Convert decimal to percentage with 2 decimal places
# {:.2%} formats the decimal as a percentage with 2 decimal places
formatted_rate = f"Success rate: {success_rate:.2%}"

# Print the formatted message
print(formatted_rate)


Success rate: 87.50%


#### **Step 5: Item Purchase Example**


In [56]:
# Define the item, price, and quantity variables
item = "book"  # Name of the item
price = 19.99  # Price of each item
quantity = 3   # Quantity purchased

# Calculate the total cost by multiplying price and quantity
total_cost = price * quantity

# Create the formatted message using f-strings
# The total cost is displayed with 2 decimal places using .2f
message = f"I bought {quantity} {item}s at £{price} each. The total cost is £{total_cost:.2f}."

# Print the message
print(message)


I bought 3 books at £19.99 each. The total cost is £59.97.





**Explanation**:
- `{quantity}` dynamically embeds the item quantity.
- `{item}s` embeds the item name and adds an "s" for pluralization.
- `{price}` inserts the price of each item.
- `{total_cost:.2f}` calculates and formats the total cost to two decimal places.

---


In [57]:
# Define variables for name, age, and address
name = "John"           # Name of the person
age = 40                # Age of the person
address = "1234 Elm Street"  # Address of the person

# Create a multiline formatted message using f-strings
# This f-string will insert variables in multiple lines
multiline_message = f"""
Name: {name}
Age: {age}
Address: {address}
"""

# Print the formatted multiline message
print(multiline_message)



Name: John
Age: 40
Address: 1234 Elm Street



# **Advanced Use Cases of F-Strings**

#### **Multiline F-Strings**




#### **Calling Functions Inside F-Strings**





In [58]:
# Define a function that returns a greeting message
def greet(name):
    return f"Hello, {name}!"  # Returns a greeting with the person's name

# Call the function inside the f-string
# The greet function is called with 'Bob' as the argument
message = f"Greetings: {greet('Bob')}"

# Print the greeting message
print(message)


Greetings: Hello, Bob!


#### **Conditional Logic Inside F-Strings**


In [59]:
# Define the temperature value
temperature = 30

# Use a conditional (ternary) operation inside the f-string
# If temperature is greater than 25, the message will say 'hot'; otherwise, it will say 'cold'
weather_message = f"The weather is {'hot' if temperature > 25 else 'cold'} today."

# Print the message
print(weather_message)


The weather is hot today.


## **Modifiers in F-Strings**

### **What are Modifiers?**
- Modifiers control the formatting of values.
- They are specified after a colon (`:`) inside the placeholder `{}`.

### **Common Modifiers**
| **Modifier**   | **Description**                                  | **Example**          | **Result**             |
|-----------------|--------------------------------------------------|----------------------|------------------------|
| `.2f`          | Fixed-point with 2 decimal places                | `{price:.2f}`        | `59.97`               |
| `.2%`          | Converts a decimal to a percentage               | `{rate:.2%}`         | `87.50%`              |
| `:<10`         | Left-align in 10 spaces                          | `{name:<10}`         | `Alice     `          |
| `:>10`         | Right-align in 10 spaces                         | `{name:>10}`         | `     Alice`          |
| `:^10`         | Center-align in 10 spaces                        | `{name:^10}`         | `   Alice   `         |

---




#### **Example: Formatting Numbers**
```python
# Define the price of the item
price = 59

# Format the price to show 2 decimal places using the f-string
# {:.2f} ensures the number is displayed with 2 decimal places
message = f"The price is {price:.2f} dollars"

# Print the formatted message
print(message)

```
**Output**:
```
The price is 59.00 dollars
```


### **Summary**

- **Placeholders** are used to represent variables or values in a string.
- **Common placeholders** include:
  - `%s` for strings.
  - `%d` for integers.
  - `%f` for floating-point numbers.
- **Modifiers** like `%.2f` control the format (e.g., number of decimal places).
- **`%r`** inserts the "representation" of an object, which is helpful for debugging and showing the exact format of objects.

---

### **Why Use F-Strings?**

- **Readable**: The syntax is concise and intuitive.
- **Dynamic**: Supports inline calculations, function calls, and conditional logic.
- **Efficient**: F-strings are faster than legacy formatting methods.

---

### **How F-Strings Work**

1. Add `f` before the string.
2. Use curly braces `{}` to include:
   - Variables.
   - Expressions (e.g., math calculations).
   - Function calls or logic.

---

### **Best Practices**

- Use **f-strings** for modern Python projects (Python 3.6+).
- Reserve `%` formatting for legacy code or specific scenarios.